google_api_lib/tasks.py

```diff
@@ -426,3 +437,172 @@ def update_meta_and_metameta_sheets_delayed(meta):
+        return
+
+    now = round(datetime.datetime.now(tz=tz.UTC).timestamp() * 1000)
+    last_update_time = cache.get(
+        "last_update_time", round(hunt.start_time.timestamp() * 1000)
+    )
+
+    action_filter = "detail.action_detail_case: EDIT"
+    time_filter = f"time > {last_update_time} AND time <= {now}"
+
+    body = {
+        "filter": f"{action_filter} AND {time_filter}",
+        "ancestorName": f"items/{hunt.settings.google_drive_folder_id}",
+    }
+
+    latest = {}
+    while True:
+        response = self.drive_activity_service().activity().query(body=body).execute()
+        activities = response.get("activities", [])
+
+        for activity in activities:
+            users = [
+                a["user"]["knownUser"]
+                for a in activity["actors"]
+                if "user" in a and "knownUser" in a["user"]
+            ]
+            drive_items = [
+                t["driveItem"] for t in activity["targets"] if "driveItem" in t
+            ]
+            for user, drive_item in itertools.product(users, drive_items):
+                user_pk = get_user_pk_from_person_name(user["personName"])
+                if not user_pk:
+                    continue
+                puzzle_pk = get_puzzle_pk_from_drive_item(drive_item["name"])
+                if not puzzle_pk:
+                    continue
+
+                timestamp = get_timestamp_from_activity(activity)
+
+                if (user_pk, puzzle_pk) not in latest:
+                    latest[(user_pk, puzzle_pk)] = timestamp
+                else:
+                    latest[(user_pk, puzzle_pk)] = max(
+                        timestamp, latest[(user_pk, puzzle_pk)]
+                    )
+
+        if "nextPageToken" not in response:
+            break
+        else:
+            body["pageToken"] = response["nextPageToken"]
+
+    with transaction.atomic():
+        user_puzzle_pairs = [
+            f"{user_pk}-{puzzle_pk}" for (user_pk, puzzle_pk) in latest.keys()
+        ]
+        old_activities = (
+            PuzzleActivity.objects.annotate(
+                user_puzzle_pair=Concat(
+                    F("user_id"), Value("-"), F("puzzle_id"), output_field=CharField()
+                )
+            )
+            .filter(user_puzzle_pair__in=user_puzzle_pairs)
+            .select_for_update()
+        )
+        old_keys = [(e.user_id, e.puzzle_id) for e in old_activities]
+
+        PuzzleActivity.objects.bulk_create(
+            [
+                PuzzleActivity(
+                    puzzle_id=puzzle_pk,
+                    user_id=user_pk,
+                    last_edit_time=timestamp,
+                )
+                for ((user_pk, puzzle_pk), timestamp) in latest.items()
+                if (user_pk, puzzle_pk) not in old_keys
+            ],
+        )
+
+        updates = []
+        for activity in old_activities:
+            activity.last_edit_time = latest[(activity.user_id, activity.puzzle_id)]
```

Comment:
is it possible for this user/puzzle pair to not exist in `latest`? won't this cause an issue?

Reply:
`old_activities` was formed from the keys of latest via the filtering in lines 574-585, so it should always exist I think? Admittedly, the filtering feels sorta hacky, but I have no idea how to make it better.